File: ios/OakApp/Features/Teams/TeamsAssistantViewModel.swift
import Foundation
import Observation

/// The team-builder assistant's view model — the SSE reducer behind the editor's
/// assistant panel (parity with web's `useTeamsAssistant` + `TeamsAssistantPanel`).
/// It holds the in-memory thread, folds a ``BuilderSSEEvent`` stream into UI state
/// one event at a time, and brokers Apply/Undo of a proposed ``TeamPatch`` against
/// the owning ``TeamEditorViewModel``'s **unsaved draft** — never the DB (the user
/// still reviews and hits Save).
///
/// `@MainActor @Observable`. It depends on the ``TeamsAssistantService`` **protocol**
/// (never `LiveTeamsAssistantService`) so it unit-tests against a fake, and holds a
/// reference to the live ``TeamEditorViewModel`` so the draft it sends each turn — and
/// the draft Apply mutates — is always the current on-screen team.
///
/// Reducer contract (`teams-assistant-sse-types.ts`, a sibling of chat MINUS `scope`):
/// `tool_activity`* → `answer_start`*/`answer_delta`* → exactly one terminal `answer`
/// (a ``BuilderAnswer``, authoritative). `answer_start` clears the streamed buffer;
/// `answer_delta` appends; the terminal answer attaches to the in-flight turn and
/// stops. A transport fault (thrown `OakError` or an in-band `error` event) drops the
/// half-finished turn, surfaces a friendly banner, and exposes a Retry — exactly like
/// the web hook.
@MainActor
@Observable
final class TeamsAssistantViewModel {

  /// One committed exchange in the panel's thread.
  struct Turn: Identifiable, Equatable {
    let id: Int
    let user: String
    /// `nil` while this turn is still streaming.
    var answer: BuilderAnswer?
  }

  /// The panel's coarse status (mirrors web's `AssistantStatus`).
  enum Status: Equatable {
    case idle
    case thinking
    case error
  }

  // MARK: Thread + streaming state

  /// The committed turns (user message + finalized ``BuilderAnswer``), oldest first.
  private(set) var turns: [Turn] = []

  /// The panel's status. `thinking` disables the composer's send.
  private(set) var status: Status = .idle

  /// The latest tool-activity label while thinking (or `nil`).
  private(set) var activity: String?

  /// The incrementally streamed `answer_markdown` for the in-flight turn.
  private(set) var streamingMarkdown: String = ""

  /// A transport-fault banner message (in-domain failures ride a normal answer).
  /// `nil` when clear.
  private(set) var errorMessage: String?

  // MARK: Apply / Undo state

  /// Turn ids whose patch has been applied to the draft (drives the "Applied ✓" label).
  private(set) var appliedTurnIds: Set<Int> = []

  /// The most recent Apply: the turn it came from and the pre-apply draft snapshot to
  /// restore on Undo. Only this turn shows an Undo affordance (mirrors web's `lastApplied`).
  private(set) var lastApplied: (turnId: Int, snapshot: TeamDraftSnapshot)?

  // MARK: Dependencies + identity

  private let service: any TeamsAssistantService
  private let editor: TeamEditorViewModel

  /// The client thread id sent as `session_id` — one in-memory conversation per mounted
  /// panel (server namespaces history under it). A remount starts a fresh thread by design.
  let sessionId: String

  /// The last message sent, kept so ``retry()`` can re-send after a recoverable failure.
  private(set) var lastMessage: String = ""

  private var nextId = 1

  /// The in-flight stream consumer; cancelled on a new turn or panel dismiss. Exposed
  /// (`private(set)`) so tests can `await streamTask?.value` to drain a scripted turn.
  private(set) var streamTask: Task<Void, Never>?

  init(service: any TeamsAssistantService, editor: TeamEditorViewModel) {
    self.service = service
    self.editor = editor
    self.sessionId = UUID().uuidString
  }

  // MARK: Derived state

  /// Whether the composer can send: not already thinking, and some non-blank text.
  func canSend(_ text: String) -> Bool {
    status != .thinking && !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
  }

  /// The one-tap first-use prompts (mirror web's `ASSISTANT_SUGGESTIONS`).
  static let suggestions = ["Check my coverage", "Fill slot 3", "Suggest an item"]

  // MARK: Actions

  /// Sends one turn: appends the user message immediately, resets the streaming state,
  /// and starts consuming the event stream with the LIVE editor draft attached. A no-op
  /// while a turn is already in flight or the text is blank.
  func send(_ message: String) {
    let text = message.trimmingCharacters(in: .whitespacesAndNewlines)
    guard status != .thinking, !text.isEmpty else { return }

    streamTask?.cancel()
    lastMessage = text
    let id = nextId
    nextId += 1

    turns.append(Turn(id: id, user: text, answer: nil))
    status = .thinking
    activity = nil
    streamingMarkdown = ""
    errorMessage = nil

    let draft = TeamsAssistantDraft(
      name: editor.name,
      format: editor.format,
      members: editor.draftWireMembers()
    )
    streamTask = Task { [weak self] in
      await self?.consume(turnId: id, message: text, draft: draft)
    }
  }

  /// Re-sends the last message after a recoverable failure (its half-finished turn was
  /// already dropped, so this re-appends cleanly). Mirrors the panel's Retry button.
  func retry() {
    guard status != .thinking, !lastMessage.isEmpty else { return }
    send(lastMessage)
  }

  /// Applies a turn's proposed patch to the editor's unsaved draft, snapshotting the
  /// pre-apply draft so ``undo()`` can restore it. Marks the turn applied.
  func apply(_ turn: Turn) {
    guard let patch = turn.answer?.teamPatch else { return }
    lastApplied = (turnId: turn.id, snapshot: editor.draftSnapshot())
    appliedTurnIds.insert(turn.id)
    editor.applyAssistantPatch(patch)
  }

  /// Restores the pre-apply draft snapshot (Undo) and un-marks the turn.
  func undo() {
    guard let applied = lastApplied else { return }
    editor.restoreDraft(applied.snapshot)
    appliedTurnIds.remove(applied.turnId)
    lastApplied = nil
  }

  /// Tears down any in-flight stream (panel dismissed / view disappeared). Committed
  /// turns are left intact — reopening the panel resumes the same in-memory thread — but
  /// a half-finished in-flight turn is dropped and the status reset to idle, so a
  /// mid-stream dismiss can never leave the panel stuck "thinking". The consumer's
  /// CancellationError path returns silently, so it won't clobber this reset.
  func cancel() {
    streamTask?.cancel()
    streamTask = nil
    guard status == .thinking else { return }
    turns.removeAll { $0.answer == nil }
    status = .idle
    activity = nil
    streamingMarkdown = ""
  }

  // MARK: Reducer

  /// Consumes one turn's event stream, folding events into UI state and committing the
  /// terminal ``BuilderAnswer`` to the turn. A thrown transport fault or a stream that
  /// ends without an answer drops the half-finished turn and raises a banner.
  private func consume(turnId: Int, message: String, draft: TeamsAssistantDraft) async {
    var terminal: BuilderAnswer?
    var inbandError: String?

    do {
      for try await event in service.send(sessionId: sessionId, message: message, draft: draft) {
        switch event {
        case let .toolActivity(_, label):
          activity = label
        case .answerStart:
          streamingMarkdown = ""
        case let .answerDelta(text):
          streamingMarkdown += text
        case let .answer(answer):
          terminal = answer
        case let .error(_, message, _):
          inbandError = message
        }
      }
    } catch is CancellationError {
      // Panel reset / unmount — drop silently (never leaves a banner).
      return
    } catch let error as OakError {
      finishWithFailure(turnId: turnId, message: TeamEditorViewModel.message(for: error))
      return
    } catch {
      finishWithFailure(turnId: turnId, message: TeamEditorViewModel.genericMessage)
      return
    }

    if let terminal {
      if let index = turns.firstIndex(where: { $0.id == turnId }) {
        turns[index].answer = terminal
      }
      status = .idle
      activity = nil
      streamingMarkdown = ""
    } else {
      finishWithFailure(
        turnId: turnId,
        message: inbandError ?? "The stream ended without an answer."
      )
    }
  }

  /// Drops the in-flight turn and raises a recoverable banner (mirrors the web hook's
  /// catch path: a retry re-sends cleanly rather than duplicating a half-turn).
  private func finishWithFailure(turnId: Int, message: String) {
    turns.removeAll { $0.id == turnId }
    errorMessage = message
    status = .error
    activity = nil
    streamingMarkdown = ""
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/features/teams/TeamsAssistantViewModel.kt
package ai.gowtam.oak.features.teams

import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.services.TeamsAssistantService
import ai.gowtam.oak.wire.BuilderAnswer
import ai.gowtam.oak.wire.BuilderSseEvent
import ai.gowtam.oak.wire.TeamsAssistantDraft
import androidx.compose.runtime.Immutable
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** One committed exchange in the panel's thread. `answer` is `null` while still streaming. */
@Immutable
data class AssistantTurn(val id: Int, val user: String, val answer: BuilderAnswer? = null)

/** The panel's coarse status (mirrors web's `AssistantStatus`). [THINKING] disables the composer's send. */
enum class AssistantStatus { IDLE, THINKING, ERROR }

/** The single renderable snapshot the Teams Assistant sheet collects. */
@Immutable
data class TeamsAssistantUiState(
    /** The committed turns (user message + finalized [BuilderAnswer]), oldest first. */
    val turns: List<AssistantTurn> = emptyList(),
    val status: AssistantStatus = AssistantStatus.IDLE,
    /** The latest tool-activity label while thinking, or `null`. */
    val activity: String? = null,
    /** The incrementally streamed `answer_markdown` for the in-flight turn. */
    val streamingMarkdown: String = "",
    /** A transport-fault banner message (in-domain failures ride a normal answer). */
    val errorMessage: String? = null,
    /** False for denylist / daily-cap refusals so the sheet hides Retry. */
    val errorIsRetryable: Boolean = true,
    /** Turn ids whose patch has been applied to the draft (drives "Applied ✓"). */
    val appliedTurnIds: Set<Int> = emptySet(),
    /** The most recent Apply's turn id — only THIS turn shows an Undo affordance. */
    val lastAppliedTurnId: Int? = null,
)

/**
 * The team-builder assistant's view model — the SSE reducer behind the editor's
 * assistant panel (history-and-teams.md D-TEAM-2; component-design.md
 * "TeamsAssistantViewModel"; mirrors iOS `TeamsAssistantViewModel` / web's
 * `useTeamsAssistant` + `TeamsAssistantPanel`). Holds the in-memory thread, folds a
 * [BuilderSseEvent] stream into UI state one event at a time, and brokers Apply/Undo of
 * a proposed `TeamPatch` against the owning [editor]'s **unsaved draft** — never the
 * DB (the user still reviews and hits Save).
 *
 * Depends on the [TeamsAssistantService] **interface** (never `LiveTeamsAssistantService`)
 * so it unit-tests against a fake, and holds a reference to the LIVE [editor] so the
 * draft it sends each turn — and the draft Apply mutates — is always the current
 * on-screen team, including manual edits made between assistant turns.
 *
 * Reducer contract (a sibling of the chat reducer MINUS `scope`): zero-or-more
 * `tool_activity`, then zero-or-more `answer_start`/`answer_delta`, then exactly one
 * terminal `answer` (a [BuilderAnswer], authoritative). A transport fault (thrown
 * [OakError] or an in-band `error` event)
 * drops the half-finished turn, surfaces a friendly banner, and exposes [retry].
 *
 * **Apply/Undo parity note:** [apply] always overwrites the "last applied" slot — a
 * SECOND apply forecloses the first one's Undo. A plain manual draft edit does NOT
 * itself clear the "last applied" pointer (this matches the shipped web
 * `TeamsAssistantPanel.tsx` / iOS behavior exactly, not merely a superficial reading of
 * the requirement copy); Undo, when offered, always restores the exact pre-apply
 * snapshot, discarding any interim edits along with the applied patch.
 */
class TeamsAssistantViewModel(
    private val service: TeamsAssistantService,
    private val editor: TeamEditorViewModel,
) : ViewModel() {

    /** The client thread id sent as `session_id` — one in-memory conversation per
     * mounted view model (server namespaces history under it). */
    val sessionId: String = UUID.randomUUID().toString()

    private val _uiState = MutableStateFlow(TeamsAssistantUiState())
    val uiState: StateFlow<TeamsAssistantUiState> = _uiState.asStateFlow()

    private var nextId = 1
    private var lastMessage: String = ""
    private var lastAppliedSnapshot: TeamDraftSnapshot? = null

    /** The in-flight stream consumer; cancelled on a new turn or [cancel]. */
    private var streamJob: Job? = null

    /** Whether the composer can send: not already thinking, and some non-blank text. */
    fun canSend(text: String): Boolean = uiState.value.status != AssistantStatus.THINKING && text.isNotBlank()

    /** Sends one turn: appends the user message immediately, resets the streaming
     * state, and starts consuming the event stream with the LIVE editor draft attached.
     * A no-op while a turn is already in flight or the text is blank. */
    fun send(message: String) {
        val text = message.trim()
        if (uiState.value.status == AssistantStatus.THINKING || text.isEmpty()) return

        streamJob?.cancel()
        lastMessage = text
        val id = nextId
        nextId += 1

        _uiState.update {
            it.copy(
                turns = it.turns + AssistantTurn(id = id, user = text),
                status = AssistantStatus.THINKING,
                activity = null,
                streamingMarkdown = "",
                errorMessage = null,
                errorIsRetryable = true,
            )
        }

        val draft = TeamsAssistantDraft(
            name = editor.uiState.value.name,
            format = editor.format,
            members = editor.draftWireMembers(),
        )
        streamJob = viewModelScope.launch { consume(id, text, draft) }
    }

    /** Re-sends the last message after a recoverable failure. Mirrors the panel's Retry button. */
    fun retry() {
        if (uiState.value.status == AssistantStatus.THINKING || lastMessage.isEmpty()) return
        send(lastMessage)
    }

    /** Applies a turn's proposed patch to the editor's unsaved draft, snapshotting the
     * pre-apply draft so [undo] can restore it. Marks the turn applied. */
    fun apply(turn: AssistantTurn) {
        val patch = turn.answer?.teamPatch ?: return
        lastAppliedSnapshot = editor.draftSnapshot()
        _uiState.update { it.copy(appliedTurnIds = it.appliedTurnIds + turn.id, lastAppliedTurnId = turn.id) }
        editor.applyAssistantPatch(patch)
    }

    /** Restores the pre-apply draft snapshot (Undo) and un-marks the turn. A no-op when
     * there is no pending "last applied" (nothing to undo, or it was already
     * foreclosed by a further apply). */
    fun undo() {
        val snapshot = lastAppliedSnapshot ?: return
        val turnId = uiState.value.lastAppliedTurnId ?: return
        editor.restoreDraft(snapshot)
        _uiState.update { it.copy(appliedTurnIds = it.appliedTurnIds - turnId, lastAppliedTurnId = null) }
        lastAppliedSnapshot = null
    }

    /** Tears down any in-flight stream (panel dismissed). Committed turns are left
     * intact — reopening resumes the same in-memory thread — but a half-finished
     * in-flight turn is dropped and the status reset to idle, so a mid-stream dismiss
     * can never leave the panel stuck "thinking". */
    fun cancel() {
        streamJob?.cancel()
        streamJob = null
        if (uiState.value.status != AssistantStatus.THINKING) return
        _uiState.update { state ->
            state.copy(
                turns = state.turns.filter { it.answer != null },
                status = AssistantStatus.IDLE,
                activity = null,
                streamingMarkdown = "",
            )
        }
    }

    override fun onCleared() {
        streamJob?.cancel()
        super.onCleared()
    }

    // ---- Reducer ----

    private suspend fun consume(turnId: Int, message: String, draft: TeamsAssistantDraft) {
        var terminal: BuilderAnswer? = null
        var inbandError: BuilderSseEvent.Error? = null

        try {
            service.send(sessionId, message, draft).collect { event ->
                when (event) {
                    is BuilderSseEvent.ToolActivity -> _uiState.update { it.copy(activity = event.label) }
                    BuilderSseEvent.AnswerStart -> _uiState.update { it.copy(streamingMarkdown = "") }
                    is BuilderSseEvent.AnswerDelta -> _uiState.update { it.copy(streamingMarkdown = it.streamingMarkdown + event.text) }
                    is BuilderSseEvent.Answer -> terminal = event.answer
                    is BuilderSseEvent.Error -> inbandError = event
                }
            }
        } catch (e: CancellationException) {
            // Panel reset / unmount — drop silently (never leaves a banner).
            throw e
        } catch (e: OakError) {
            finishWithFailure(
                turnId,
                TeamEditorViewModel.message(e),
                isRetryable = e.isRetryableAgentTurn(),
            )
            return
        } catch (e: Exception) {
            finishWithFailure(turnId, TeamEditorViewModel.GENERIC_MESSAGE)
            return
        }

        val answer = terminal
        if (answer != null) {
            _uiState.update { state ->
                state.copy(
                    turns = state.turns.map { if (it.id == turnId) it.copy(answer = answer) else it },
                    status = AssistantStatus.IDLE,
                    activity = null,
                    streamingMarkdown = "",
                )
            }
        } else {
            val inband = inbandError
            finishWithFailure(
                turnId,
                inband?.message ?: "The stream ended without an answer.",
                isRetryable = inband == null || !OakError.isSpendControlRefusal(inband.code),
            )
        }
    }

    /** Drops the in-flight turn and raises a banner (a retry re-sends
     * cleanly rather than duplicating a half-turn). Spend-control refusals
     * hide Retry ([errorIsRetryable] false). */
    private fun finishWithFailure(turnId: Int, message: String, isRetryable: Boolean = true) {
        _uiState.update {
            it.copy(
                turns = it.turns.filterNot { turn -> turn.id == turnId },
                errorMessage = message,
                errorIsRetryable = isRetryable,
                status = AssistantStatus.ERROR,
                activity = null,
                streamingMarkdown = "",
            )
        }
    }

    private fun OakError.isRetryableAgentTurn(): Boolean = when (this) {
        is OakError.Http -> !OakError.isSpendControlRefusal(code)
        else -> true
    }

    companion object {
        /** The one-tap first-use prompts (mirrors web's `ASSISTANT_SUGGESTIONS`). */
        val suggestions = listOf("Check my coverage", "Fill slot 3", "Suggest an item")
    }
}
